README.md
# UnitfulAstro.jl

UnitfulAstro.jl is an extension of [Unitful.jl](https://github.com/ajkeller34/Unitful.jl) to include
units commonly encountered in astronomy.

[![Build Status](https://travis-ci.org/JuliaAstro/UnitfulAstro.jl.svg?branch=master)](https://travis-ci.org/JuliaAstro/UnitfulAstro.jl)
[![Build Status](https://ci.appveyor.com/api/projects/status/9tgicxbo6ud533hp/branch/master?svg=true)](https://ci.appveyor.com/project/mweastwood/unitfulastro-jl/branch/master)

**Author:** Michael Eastwood

**License:** MIT "Expat"

**Documentation:** http://juliaastro.github.io/UnitfulAstro.jl/stable

## Summary of Units

The list of additional units is given below:

* [Erg](https://en.wikipedia.org/wiki/Erg)
* [Dyne](https://en.wikipedia.org/wiki/Dyne)
* [Arc minute](https://en.wikipedia.org/wiki/Minute_and_second_of_arc)
* [Arc second](https://en.wikipedia.org/wiki/Minute_and_second_of_arc)
* [Julian year](https://en.wikipedia.org/wiki/Julian_year_%28astronomy%29)
* [Astronomical unit](https://en.wikipedia.org/wiki/Astronomical_unit)
* [Light year](https://en.wikipedia.org/wiki/Light-year)
* [Parsec](https://en.wikipedia.org/wiki/Parsec)
* [Jansky](https://en.wikipedia.org/wiki/Jansky)
* [Angstrom](https://en.wikipedia.org/wiki/%C3%85ngstr%C3%B6m)
* [Solar radius](https://en.wikipedia.org/wiki/Solar_radius)
* [Solar irradiance](https://en.wikipedia.org/wiki/Solar_irradiance)
* [Solar luminosity](https://en.wikipedia.org/wiki/Solar_luminosity)
* [Solar mass](https://en.wikipedia.org/wiki/Solar_mass) [1]
* [Earth radius](https://en.wikipedia.org/wiki/Earth_radius) [2]
* [Earth mass](https://en.wikipedia.org/wiki/Earth_mass) [1]
* [Jupiter radius](https://en.wikipedia.org/wiki/Jupiter_radius) [2]
* [Jupiter mass](https://en.wikipedia.org/wiki/Jupiter_mass) [1]
* [Fifty-one ergs](https://en.wikipedia.org/wiki/Foe_%28unit%29)
* [Solar flux unit](https://en.wikipedia.org/wiki/Solar_flux_unit)
* [Total electron content unit](http://www.swpc.noaa.gov/phenomena/total-electron-content)

In addition the following magnitudes are supported:

* [Absolute and apparent bolometric magnitude](https://en.wikipedia.org/wiki/Absolute_magnitude#Bolometric_magnitude)
* [AB magnitude](https://en.wikipedia.org/wiki/AB_magnitude)
* [Johnson-Cousins UBVRI magnitudes](https://en.wikipedia.org/wiki/UBV_photometric_system)
* [Gunn griz magnitudes](https://www.cfa.harvard.edu/~dfabricant/huchra/ay145/mags.html#definitions)

Where applicable, all units are defined by the IAU standard.

[1]: gravitational constant times mass also given

[2]: polar and equatorial radii


NEWS.md
# News

## v0.2.0

*2018-11-03*

* Require Julia 0.7.

## v0.1.1

*2018-02-14*

* Fix a warning created by the addition of some cgs units to default Unitful.jl.

## v0.1.0

*2017-09-27*

* Initial tag.
* List of units: erg, dyn, arcmin, arcsec, Julian year, AU, light year, parsec, Jansky,
  solar radius, solar irradiance, solar luminosity, solar mass, Earth radius, Earth mass, Jupiter
  radius, Jupiter mass, solar flux unit, total electron content unit.



test/runtests.jl
using UnitfulAstro
using Test

const u = UnitfulAstro

@testset "UnitfulAstro.jl" begin
    @testset "unit tests" begin
        @test 10^7*u.erg == 1*u.J
        @test 10^5*u.dyn == 1*u.N
        @test 60*u.arcminute == 1*u.°
        @test 60*u.arcsecond == 1*u.arcminute
        @test 3600*u.arcsecond == 1*u.°
        @test 365.25*u.d == 1*u.yr
        @test 1*u.ly == 1*u.c*u.yr
        @test 1*u.pc == 1*u.AU/u.arcsecond
        @test 1*u.SFU ≈ 10000*u.Jy
        @test 1*u.TECU == 1e12*u.cm^-2
    end

    @testset "simple calculations" begin
        # gravitational acceleration at Earth's surface
        @test isapprox(1*u.GMearth/u.Rearth^2, 1*u.gn, atol=0.001*u.gn)
        # Solar flux at Earth's surface
        @test isapprox(1*u.Lsun/(4π*u.AU^2), 1*u.Ssun, atol=0.001*u.Ssun)
        # rest mass energy of the sun
        @test isapprox(1*u.Msun*u.c^2, 1787*u.foe, atol=0.5*u.foe)
    end

    @testset "magnitudes" begin
        @test isapprox(3631*u.Jy, 0*u.AB_mag, atol=0.001*u.AB_mag)
        @test isapprox(36.31*u.Jy, 5*u.AB_mag, atol=0.001*u.AB_mag)
        @test isapprox(363.1*u.mJy, 10*u.AB_mag, atol=0.001*u.AB_mag)
        @test isapprox(3.631*u.mJy, 15*u.AB_mag, atol=0.001*u.AB_mag)
        @test isapprox(5*u.AB_mag + 5*u.AB_mag,  4.247425010840047*u.AB_mag, atol=0.001*u.AB_mag)
        @test isapprox(5*u.AB_mag / 100, 10*u.AB_mag, atol=0.001*u.AB_mag)
        @test isapprox(5*u.AB_mag + 10*u.Jy, 46.31*u.Jy, atol=0.001*u.AB_mag)

        @test isapprox(3.0128e28*u.W, 1u.bol_Mag, atol=0.001*u.bol_Mag)
        @test isapprox(2.518_021_002e-8*u.W*u.m^-2, 1u.bol_mag, atol=0.001*u.bol_mag)

        @test isapprox(1u.U_mag, 1810*u.Jy, atol=0.001*u.U_mag)
        @test isapprox(1u.B_mag, 4260*u.Jy, atol=0.001*u.B_mag)
        @test isapprox(1u.V_mag, 3640*u.Jy, atol=0.001*u.V_mag)
        @test isapprox(1u.R_mag, 3080*u.Jy, atol=0.001*u.R_mag)
        @test isapprox(1u.I_mag, 2550*u.Jy, atol=0.001*u.I_mag)
        @test isapprox(1u.J_mag, 1600*u.Jy, atol=0.001*u.J_mag)
        @test isapprox(1u.H_mag, 1080*u.Jy, atol=0.001*u.H_mag)
        @test isapprox(1u.K_mag,  670*u.Jy, atol=0.001*u.K_mag)
        @test isapprox(1u.g_mag, 3730*u.Jy, atol=0.001*u.g_mag)
        @test isapprox(1u.r_mag, 4490*u.Jy, atol=0.001*u.r_mag)
        @test isapprox(1u.i_mag, 4760*u.Jy, atol=0.001*u.i_mag)
        @test isapprox(1u.z_mag, 4810*u.Jy, atol=0.001*u.z_mag)

        @test iszero(1u.B_mag - 1u.V_mag)
        @test isapprox(5*u.B_mag - 1u.V_mag, 4)
        @test isapprox(1u.B_mag - 5*u.V_mag, -4)
        @test isapprox(1u.B_mag - 0.5u.V_mag, 0.5)
        @test_throws u.MagnitudeError 1u.B_mag + 2u.V_mag
        @test_throws u.MagnitudeError 1u.B_mag * 2u.V_mag
        @test_throws u.MagnitudeError 1u.B_mag / 2u.V_mag
        @test_throws u.MagnitudeError 1u.B_mag // 2u.V_mag

        # Extra tests for rootpower calculations
        flux = 10u.erg
        mag = 1u.mag
        @test isapprox(flux * mag, 3.981071705534972u.erg)
        flux = 10u.erg/u.s
        @test isapprox(flux * mag, 3.981071705534972u.erg/u.s)
        flux = 10u.erg/u.s/u.cm^2
        @test isapprox(flux * mag, 3.981071705534972u.erg/u.s/u.cm^2)
    end
end



latest/search_index.js
var documenterSearchIndex = {"docs": [

{
    "location": "#",
    "page": "Home",
    "title": "Home",
    "category": "page",
    "text": ""
},

{
    "location": "#UnitfulAstro.jl-Documentation-1",
    "page": "Home",
    "title": "UnitfulAstro.jl Documentation",
    "category": "section",
    "text": "UnitfulAstro.jl is an extension of Unitful.jl to include units commonly encountered in astronomy."
},

{
    "location": "#Table-of-Units-1",
    "page": "Home",
    "title": "Table of Units",
    "category": "section",
    "text": "The list of additional units is given below:Name Binding Value\nErg UnitfulAstro.erg 1//10000000 J\nDyne UnitfulAstro.dyn 1//100000 N\nArcminute UnitfulAstro.arcminute 0.016666666666666666 °\nArcsecond UnitfulAstro.arcsecond 0.0002777777777777778 °\nJulian year UnitfulAstro.yr 365.25 dy\nAstronomical unit UnitfulAstro.AU 1.495978707e13 cm\nLight year UnitfulAstro.ly 9.4607304725808e17 cm\nParsec UnitfulAstro.pc 3.0856775814913674e18 cm\nJansky UnitfulAstro.Jy 1.0e-23 erg Hz^-1 cm^-2 s^-1\nSolar radius UnitfulAstro.Rsun 6.957e10 cm\nSolar irradiance UnitfulAstro.Ssun 1.361e6 erg cm^-2 s^-1\nSolar luminosity UnitfulAstro.Lsun 3.828e33 erg s^-1\nG × Solar mass UnitfulAstro.GMsun 1.3271244e20 m^3 s^-2\nSolar mass UnitfulAstro.Msun 1.9884754153381438e33 g\nEarth radius (equatorial) UnitfulAstro.Rearth_e 6.3781e8 cm\nEarth radius (polar) UnitfulAstro.Rearth_p 6.3568e8 cm\nEarth radius UnitfulAstro.Rearth 6.3781e8 cm\nG × Earth mass UnitfulAstro.GMearth 3.986004e14 m^3 s^-2\nEarth mass UnitfulAstro.Mearth 5.972364730419773e27 g\nJupiter radius (equatorial) UnitfulAstro.Rjup_e 7.1492e9 cm\nJupiter radius (polar) UnitfulAstro.Rjup_p 6.6854e9 cm\nJupiter radius UnitfulAstro.Rjup 7.1492e9 cm\nG × Jupiter mass UnitfulAstro.GMjup 1.2668653e17 m^3 s^-2\nJupiter mass UnitfulAstro.Mjup 1.8981871658715508e30 g\nSolar flux unit UnitfulAstro.SFU 10//1 kJy\nTotal electron content unit UnitfulAstro.TECU 1.0e12 cm^-2"
},

{
    "location": "#Examples-1",
    "page": "Home",
    "title": "Examples",
    "category": "section",
    "text": "julia> using Unitful, UnitfulAstro\n\njulia> uconvert(u\"erg\", 1 * Unitful.kg * Unitful.gn * Unitful.m)\n9.80665e7 erg\n\njulia> uconvert(u\"Jy\", 1.23e-20 * u\"erg/s/cm^2/Hz\")\n1230.0000000000002 Jy\n\njulia> uconvert(u\"ly\", 1 * u\"pc\")\n3.2615637771674337 ly"
},

{
    "location": "#IAU-Resolutions-1",
    "page": "Home",
    "title": "IAU Resolutions",
    "category": "section",
    "text": "Copies of recent IAU resolutions which formalize the definitions of some units used in this package are linked below.IAU 2012 (pdf)\nIAU 2015 (pdf)"
},

]}


src/UnitfulAstro.jl
__precompile__()

module UnitfulAstro

import Unitful
using Unitful: @unit, @logscale, @logunit

macro import_from_unitful(args...)
    expr = Expr(:block)
    for arg in args
        use_SI_prefixes, sym = should_we_use_SI_prefixes(arg)
        if use_SI_prefixes
            for prefix in Unitful.si_prefixes
                sym′ = Symbol(prefix, sym)
                push!(expr.args, :(import Unitful.$sym′))
            end
        else
            push!(expr.args, :(import Unitful.$sym))
        end
    end
    expr
end

function should_we_use_SI_prefixes(arg::Expr)
    if arg.head == :(call)
        if arg.args[1] == :(~)
            return true, arg.args[2]
        else
            error("incorrect first argument")
        end
    else
        error("incorrect expression head")
    end
end
should_we_use_SI_prefixes(arg::Symbol) = false, arg

@import_from_unitful ~m ~s ~A ~K ~cd
@import_from_unitful ~L ~Hz ~N ~Pa ~J ~W ~C ~V ~Ω ~S ~F ~H ~T ~Wb ~lm ~lx ~Bq ~Gy ~Sv ~kat ~eV
@import_from_unitful sr rad ° °C °F Ra minute hr d wk ~bar atm Torr
@import_from_unitful q c0 c μ0 µ0 ε0 ϵ0 Z0 G gn ge h ħ Φ0 me mn mp μB µB Na R k σ
@import_from_unitful inch ft yd mi ac lb oz dr gr lbf

# CGS units have been moved to default Unitful.jl now
@import_from_unitful dyn erg

import UnitfulAngles: arcminute, arcsecond

@unit yr         "yr"       JulianYear                365.25*d                  true
@unit AU         "AU"       AstronomicalUnit          149_597_870_700.0*m       false  # cf IAU 2012
@unit ly         "ly"       LightYear                 1*c*yr                    false
@unit pc         "pc"       Parsec                    1*AU/arcsecond            true   # cf IAU 2015
# Note that IAU 2015 defines the small angle approximation as exact in the definition of the parsec.
@unit Jy         "Jy"       Jansky                    1e-23erg*s^-1*cm^-2*Hz^-1 true
# Note that Jy uses a Float64 conversion factor because 10^23 overflows Int64.
@unit angstrom   "Å"        Angstrom                  1e-10*m                   false

# SOLAR CONVERSION CONSTANTS (IAU 2015)
@unit Rsun       "R⊙"       SolarRadius               6.957e8*m                 false
@unit Ssun       "S⊙"       SolarIrradiance           1361.0*W*m^-2             false
@unit Lsun       "L⊙"       SolarLuminosity           3.828e26*W                false
# TODO solar effective temperature
@unit GMsun      "GM⊙"      GSolarMass                1.327_124_4e20*m^3*s^-2   false
@unit Msun       "M⊙"       SolarMass                 1*GMsun/G                 false

# PLANETARY CONVERSION CONSTANTS (IAU 2015)
@unit Rearth_e   "R⊕ₑ"      EarthEquatorialRadius     6.3781e6*m                false
@unit Rearth_p   "R⊕ₚ"      EarthPolarRadius          6.3568e6*m                false
@unit Rjup_e     "Rjupₑ"    JupiterEquatorialRadius   7.1492e7*m                false
@unit Rjup_p     "Rjupₚ"    JupiterPolarRadius        6.6854e7*m                false
@unit GMearth    "GM⊕"      GEarthMass                3.986_004e14*m^3*s^-2     false
@unit GMjup      "GMjup"    GJupiterMass              1.266_865_3e17*m^3*s^-2   false
@unit Mearth     "M⊕"       EarthMass                 1*GMearth/G               false
@unit Mjup       "Mjup"     JupiterMass               1*GMjup/G                 false
@unit Rearth     "R⊕"       EarthRadius               1*Rearth_e                false
@unit Rjup       "Rjup"     JupiterRadius             1*Rjup_e                  false
# Note that IAU 2015 states that when the radius is not specified as polar or equatorial, the
# equatorial radius should be used.

# "ten to the Fifty-One Ergs" (approximate energy of a supernova explosition)
@unit foe        "foe"      FiftyOneErgs              1e51*erg                  false

# Solar flux unit
# https://en.wikipedia.org/wiki/Solar_flux_unit
@unit SFU        "SFU"      SolarFluxUnit             10*kJy                    false

# Total electron content unit (used for ionospheric physics and low-frequency radio astronomy)
@unit TECU       "TECU"     TotalElectronContentUnit  1e16*m^-2                 false

@logscale mag "mag" Magnitude 10 -2.5 false

#https://en.wikipedia.org/wiki/AB_magnitude     
@logunit    AB_mag  "AB mag"        Magnitude   3631Jy

# Bolometric magnitudes (cf IAU 2015)
# https://en.wikipedia.org/wiki/Absolute_magnitude#Bolometric_magnitude
@logunit    bol_Mag "bol Mag"       Magnitude   3.0128e28*W
@logunit    bol_mag "bol mag"       Magnitude   2.518_021_002e-8*W*m^-2

#https://en.wikipedia.org/wiki/Apparent_magnitude#Standard_reference_values
#https://www.cfa.harvard.edu/~dfabricant/huchra/ay145/mags.html
@logunit    U_mag   "Johnson U mag" Magnitude   1810Jy
@logunit    B_mag   "Johnson B mag" Magnitude   4260Jy
@logunit    V_mag   "Johnson V mag" Magnitude   3640Jy
@logunit    R_mag   "Johnson R mag" Magnitude   3080Jy
@logunit    I_mag   "Johnson I mag" Magnitude   2550Jy
@logunit    J_mag   "Johnson J mag" Magnitude   1600Jy
@logunit    H_mag   "Johnson H mag" Magnitude   1080Jy
@logunit    K_mag   "Johnson K mag" Magnitude    670Jy
@logunit    g_mag   "Gunn g mag"    Magnitude   3730Jy
@logunit    r_mag   "Gunn r mag"    Magnitude   4490Jy
@logunit    i_mag   "Gunn i mag"    Magnitude   4760Jy
@logunit    z_mag   "Gunn z mag"    Magnitude   4810Jy


struct MagnitudeError <: Exception
    x
end
Base.showerror(io::IO, e::MagnitudeError) = print(io, "MagnitudeError: ", e.x)

# https://en.wikipedia.org/wiki/Color_index
# for color indices (subtracting magnitudes of different passbands)
import Base.-
function -(x::Unitful.Level{L,S,T},y::Unitful.Level{L,S,T}) where {L<:Magnitude,S,T<:Unitful.Quantity}
    return Unitful.Level{L,S}(-(x.val, y.val))
end
import Base.-
function -(x::Unitful.Level{L,S,T},y::Unitful.Level{L,R,T}) where {L<:Magnitude,S,R,T<:Unitful.Quantity}
    return Unitful.ustrip(x) - Unitful.ustrip(y)
end
import Base.+
function +(x::Unitful.Level{L,S,T},y::Unitful.Level{L,S,T}) where {L<:Magnitude,S,T<:Unitful.Quantity}
    return Unitful.Level{L,S}(+(x.val, y.val))
end
import Base.+
function +(x::Unitful.Level{L,S,T},y::Unitful.Level{L,R,T}) where {L<:Magnitude,S,R,T<:Unitful.Quantity}
    throw(MagnitudeError("an invalid operation was attempted with magnitudes: $x, $y"))
end
import Base.*
function *(x::Unitful.Level{L,S,T},y::Unitful.Level{L,R,T}) where {L<:Magnitude,S,R,T<:Unitful.Quantity}
    throw(MagnitudeError("an invalid operation was attempted with magnitudes: $x, $y"))
end
import Base./
function /(x::Unitful.Level{L,S,T},y::Unitful.Level{L,R,T}) where {L<:Magnitude,S,R,T<:Unitful.Quantity}
    throw(MagnitudeError("an invalid operation was attempted with magnitudes: $x, $y"))
end
import Base.//
function //(x::Unitful.Level{L,S,T},y::Unitful.Level{L,R,T}) where {L<:Magnitude,S,R,T<:Unitful.Quantity}
    throw(MagnitudeError("an invalid operation was attempted with magnitudes: $x, $y"))
end


const localunits = Unitful.basefactors
function __init__()
    merge!(Unitful.basefactors, localunits)
    Unitful.register(UnitfulAstro)
end

end
